penty/pentypes/builtins.py

```diff
@@ -291,6 +291,136 @@ def float_str(self_ty):
+    if not issubclass(self_ty, complex):
+        raise TypeError
+    return complex
+
+def complex_bool(self_ty):
+    if not issubclass(self_ty, complex):
+        raise TypeError
+    return bool
+
+def complex_divmod(self_ty, other_ty):
+    if not issubclass(self_ty, complex):
+        raise TypeError
+    raise TypeError
+
+def complex_complex(self_ty):
+    if not issubclass(self_ty, complex):
+        raise TypeError
+    return complex
+
+def complex_init(real_ty, imag_ty=None):
+    from penty.penty import Types
+    real_ty = _astype(real_ty)
+
+    if imag_ty is None:
+        if issubclass(real_ty, complex):
+            return complex
+        Types[real_ty]['__float__'](real_ty)
+        return complex
+
+    # interestingly, if the real part is a complex, everything is fine (!)
+    imag_ty = _astype(imag_ty)
+    if issubclass(real_ty, complex):
+        if issubclass(imag_ty, complex):
+            return complex
+        Types[imag_ty]['__float__'](imag_ty)
+        return complex
+
+    Types[real_ty]['__float__'](real_ty)
+    if issubclass(imag_ty, complex):
+        return complex
+    Types[imag_ty]['__float__'](imag_ty)
+    return complex
+
+def complex_eq(self_ty, other_ty):
+    if not issubclass(self_ty, complex):
+        raise TypeError
+    return bool
+
+def complex_ne(self_ty, other_ty):
+    if not issubclass(self_ty, complex):
+        raise TypeError
+    return bool
+
+def complex_int(self_ty):
+    if not issubclass(self_ty, complex):
+        raise TypeError
+    raise TypeError
+
+def complex_make_binop(operator):
+    def binop(self_ty, other_ty):
+        self_ty, other_ty = _astype(self_ty), _astype(other_ty)
+        if not issubclass(self_ty, complex):
+            raise TypeError
+
+        if not issubclass(other_ty, (int, float, complex)):
+            raise TypeError
+
+        return complex
+
+    return _CFT[binop, operator]
+
+def complex_make_unaryop(operator):
+    def unaryop(self_ty):
+        if not issubclass(self_ty, complex):
+            raise TypeError
+        return complex
+    return _CFT[unaryop, operator]
+
+
+def complex_make_boolop(operator):
+    def boolop(self_ty, other_ty):
+        if not issubclass(self_ty, complex):
+            raise TypeError
+        raise TypeError
+    return _CFT[boolop, operator]
+
+def complex_make_biniop(operator):
+    def biniop(self_ty, other_ty):
+        result_ty = operator(self_ty, other_ty)
+        # complex are immutable so we don't update self_ty
+        return result_ty
+    return _CFT[biniop, operator]
+
+def complex_str(self_ty):
+    if not issubclass(self_ty, complex):
+        raise TypeError
+    return str
+
+_complex_attrs = {
+    '__abs__': _CFT[complex_abs, complex.__abs__],
+    '__add__': complex_make_binop(_operator.add),
+    '__bool__': _CFT[complex_bool, complex.__bool__],
+    '__divmod__': _FT[complex_divmod],
+    '__eq__': _CFT[complex_eq, _operator.eq],
+    '__complex__': _CFT[complex_complex, complex],
```

Comment:
`'__complex__' in dir(complex)` yields `False`. Did I miss something?

Reply:
Correct!